File: src/lib/drc.js
import { TISSUES, MATERIALS } from '../data/bioData';

// ============================================================
// Debye Bio — Dynamic Design Rule Check Engine
// Runs against actual canvas nodes & connections, not scenarios
// ============================================================

/**
 * Resolve tissue from node label
 */
function findTissue(label) {
  if (!label) return null;
  const l = label.toLowerCase();
  for (const [key, t] of Object.entries(TISSUES)) {
    if (l.includes(key) || l.includes(t.name.toLowerCase().split(' ')[0])) return { key, ...t };
  }
  if (l.includes('subcutaneous') || l.includes('glucose')) return { key: 'subcutaneous', ...TISSUES.subcutaneous };
  if (l.includes('cardiac') || l.includes('heart')) return { key: 'cardiac', ...TISSUES.cardiac };
  if (l.includes('brain') || l.includes('cortex')) return { key: 'cortical', ...TISSUES.cortical };
  if (l.includes('blood')) return { key: 'blood', ...TISSUES.blood };
  if (l.includes('skin') || l.includes('epidermis')) return { key: 'skin_epidermis', ...TISSUES.skin_epidermis };
  if (l.includes('nerve')) return { key: 'peripheral_nerve', ...TISSUES.peripheral_nerve };
  if (l.includes('gastric') || l.includes('stomach')) return { key: 'gastric', ...TISSUES.gastric };
  if (l.includes('wound')) return { key: 'wound', ...TISSUES.wound };
  if (l.includes('retina')) return { key: 'retina', ...TISSUES.retina };
  if (l.includes('bone')) return { key: 'bone', ...TISSUES.bone };
  return null;
}

/**
 * Resolve material from node label
 */
function findMaterial(label) {
  if (!label) return null;
  const l = label.toLowerCase();
  for (const [key, m] of Object.entries(MATERIALS)) {
    if (l.includes(key) || l.includes(m.name.toLowerCase().split(' ')[0])) return { key, ...m };
  }
  if (l.includes('pedot') || l.includes('pss')) return { key: 'pedot', ...MATERIALS.pedot };
  if (l.includes('pt-ir') || l.includes('platinum-iridium')) return { key: 'platinum_iridium', ...MATERIALS.platinum_iridium };
  if (l.includes('platinum') || l.includes('pt')) return { key: 'platinum', ...MATERIALS.platinum };
  if (l.includes('gold') || l.includes('au')) return { key: 'gold', ...MATERIALS.gold };
  if (l.includes('irox') || l.includes('iridium oxide')) return { key: 'iridium_oxide', ...MATERIALS.iridium_oxide };
  if (l.includes('graphene')) return { key: 'graphene', ...MATERIALS.graphene };
  if (l.includes('cnt') || l.includes('carbon nanotube')) return { key: 'carbon_nanotube', ...MATERIALS.carbon_nanotube };
  return null;
}

/**
 * Main DRC runner — analyses actual canvas nodes and edges
 */
export function runDRC(nodes, scenarioId, edges = []) {
  const errors = [];
  const warnings = [];
  let passed = 0;

  if (nodes.length === 0) {
    return { errors, warnings, passed: 0 };
  }

  // ---- Classify nodes ----
  const tissueNodes = [];
  const electrodeNodes = [];
  const materialNodes = [];
  const electronicsNodes = [];

  for (const node of nodes) {
    const d = node.data || {};
    const label = (d.label || '').toLowerCase();

    if (node.type === 'biology' || d.type === 'tissue') {
      const t = findTissue(d.label);
      tissueNodes.push({ node, tissue: t });
    } else if (node.type === 'material') {
      const m = findMaterial(d.label);
      materialNodes.push({ node, material: m });
    } else if (node.type === 'electronics') {
      electronicsNodes.push(node);
      if (label.includes('electrode')) {
        const m = findMaterial(d.material || d.label);
        electrodeNodes.push({ node, material: m, area: d.area || 1000 });
      }
    }
  }

  // ---- RULE: Check electrode impedance range ----
  for (const elec of electrodeNodes) {
    if (elec.material && elec.material.eis_factor > 2.0 && elec.area < 1500) {
      errors.push({
        id: "BIO-001",
        title: "ELECTRODE IMPEDANCE OUT OF RANGE",
        affected: elec.node.data.label,
        message: `Electrode "${elec.node.data.label}" (${elec.area} µm², ${elec.material.name}) has high impedance at 1 kHz. For reliable sensing, increase area to ≥1500 µm² or switch to a lower-impedance material like PEDOT:PSS or IrOx.`,
        fixable: true,
      });
    } else {
      passed++;
    }
  }

  // ---- RULE: ISO 10993 biocompatibility ----
  for (const mat of materialNodes) {
    if (mat.material && !mat.material.iso10993) {
      errors.push({
        id: "BIO-006",
        title: "MATERIAL NOT ISO 10993 APPROVED",
        affected: mat.node.data.label,
        message: `"${mat.material.name}" has not passed ISO 10993 biocompatibility testing. Not approved for implantable or in-vivo use. Suitable for research/in-vitro only.`,
        fixable: false,
      });
    } else if (mat.material) {
      passed++;
    }
  }

  // ---- RULE: Chronic implant without encapsulant ----
  const hasChronicTissue = tissueNodes.some(t =>
    t.tissue && (t.tissue.key === 'cortical' || t.tissue.key === 'peripheral_nerve' ||
      t.tissue.key === 'retina' || t.tissue.key === 'gastric')
  );
  const hasEncapsulant = materialNodes.some(m =>
    m.material && (m.material.key === 'parylene_c' || m.material.key === 'titanium')
  );

  if (hasChronicTissue && !hasEncapsulant) {
    errors.push({
      id: "BIO-007",
      title: "NO ENCAPSULANT ON CHRONIC IMPLANT",
      affected: "Device Assembly",
      message: `Design targets chronic implant tissue but has no encapsulant (Parylene-C) or structural housing (Titanium). Chronic implants require hermetic sealing to prevent moisture ingress and device failure.`,
      fixable: false,
    });
  } else if (hasChronicTissue && hasEncapsulant) {
    passed++;
  }

  // ---- RULE: Charge injection limit ----
  for (const elec of electrodeNodes) {
    const pulseGen = electronicsNodes.find(n =>
      (n.data.label || '').toLowerCase().includes('pulse') ||
      (n.data.label || '').toLowerCase().includes('stimulat')
    );
    if (pulseGen && elec.material) {
      const current_mA = parseFloat(pulseGen.data?.current) || 2;
      const pulseWidth_ms = parseFloat(pulseGen.data?.pulseWidth) || 1;
      const charge_uC = current_mA * pulseWidth_ms; // µC per phase
      const area_cm2 = (elec.area || 1000) * 1e-8;
      const chargeDensity_mC = (charge_uC * 1e-3) / area_cm2; // mC/cm²

      if (chargeDensity_mC > elec.material.cil) {
        errors.push({
          id: "BIO-003",
          title: "CHARGE INJECTION LIMIT EXCEEDED",
          affected: elec.node.data.label,
          message: `Charge density ${chargeDensity_mC.toFixed(1)} mC/cm² exceeds the safe limit for ${elec.material.name} (${elec.material.cil} mC/cm²). Risk of tissue damage and electrode dissolution. Reduce current, shorten pulse width, or increase electrode area.`,
          fixable: false,
        });
      } else {
        passed++;
      }
    }
  }

  // ---- RULE: No reference electrode ----
  const hasRefElectrode = nodes.some(n =>
    (n.data?.label || '').toLowerCase().includes('reference')
  );
  const hasPotentiostat = nodes.some(n =>
    (n.data?.label || '').toLowerCase().includes('potentiostat') ||
    (n.data?.label || '').toLowerCase().includes('adc')
  );
  if (hasPotentiostat && !hasRefElectrode && electrodeNodes.length > 0) {
    warnings.push({
      id: "BIO-004",
      title: "NO REFERENCE ELECTRODE",
      affected: "Measurement System",
      message: `Design includes measurement electronics but no reference electrode. A stable reference (e.g., Ag/AgCl) is critical for drift-free potentiometric and amperometric measurements.`,
      fixable: false,
    });
  } else if (hasPotentiostat && hasRefElectrode) {
    passed++;
  }

  // ---- RULE: Material fouling for chronic ----
  for (const elec of electrodeNodes) {
    if (elec.material && !elec.material.chronic_safe && hasChronicTissue) {
      warnings.push({
        id: "WARN-007",
        title: "MATERIAL NOT RATED FOR CHRONIC USE",
        affected: elec.node.data.label,
        message: `${elec.material.name} is rated for max ${elec.material.max_years} year(s) implantation. Your design targets chronic tissue. Expect impedance increase of 15–40% due to protein fouling and fibrous encapsulation over time.`,
        fixable: false,
      });
    } else if (elec.material && elec.material.chronic_safe) {
      passed++;
    }
  }

  // ---- RULE: Motion artifact on wearable tissues ----
  const hasWearableTissue = tissueNodes.some(t =>
    t.tissue && (t.tissue.key === 'skin_epidermis' || t.tissue.key === 'subcutaneous')
  );
  const hasMotionFilter = nodes.some(n =>
    (n.data?.label || '').toLowerCase().includes('filter') ||
    (n.data?.label || '').toLowerCase().includes('motion')
  );
  if (hasWearableTissue && !hasMotionFilter && electrodeNodes.length > 0) {
    warnings.push({
      id: "WARN-003",
      title: "NO MOTION ARTIFACT FILTER",
      affected: "Signal Chain",
      message: `Design is placed on wearable tissue (skin/subcutaneous) but includes no motion artifact rejection. For ambulatory use, add an adaptive filter or accelerometer-based artifact cancellation.`,
      fixable: false,
    });
  } else if (hasWearableTissue && hasMotionFilter) {
    passed++;
  }

  // ---- RULE: No temperature compensation ----
  const hasTempComp = nodes.some(n =>
    (n.data?.label || '').toLowerCase().includes('temp') ||
    (n.data?.label || '').toLowerCase().includes('thermistor')
  );
  if (electrodeNodes.length > 0 && !hasTempComp) {
    warnings.push({
      id: "WARN-011",
      title: "NO TEMPERATURE COMPENSATION",
      affected: "Measurement Accuracy",
      message: `Electrode impedance drifts ±2% per °C. Without a temperature sensor in the signal chain, measurement accuracy will degrade in environments with temperature variation (e.g., wearables, point-of-care).`,
      fixable: false,
    });
  } else if (hasTempComp) {
    passed++;
  }

  // ---- RULE: No MCU/Wireless for complex designs ----
  const hasMCU = nodes.some(n => {
    const l = (n.data?.label || '').toLowerCase();
    return l.includes('mcu') || l.includes('microcontroller') || l.includes('wireless');
  });
  if (nodes.length > 4 && !hasMCU) {
    warnings.push({
      id: "WARN-020",
      title: "NO DATA PROCESSING UNIT",
      affected: "System Architecture",
      message: `Complex design (${nodes.length} components) but no microcontroller or data processing unit. Consider adding an MCU for signal processing, data logging, or wireless transmission.`,
      fixable: false,
    });
  } else if (hasMCU) {
    passed++;
  }

  // ---- Base pass count for general topology checks ----
  if (tissueNodes.length > 0) passed++;
  if (electrodeNodes.length > 0) passed++;
  if (edges && edges.length > 0) passed++;

  return { errors, warnings, passed };
}

File: pkg/debye/drc.go
package debye

import (
	"fmt"
	"math"
	"strings"
)

func RunDRC(req Request) DRCResponse {
	design := NormalizeDesign(req)
	var errors []Violation
	var warnings []Violation
	passed := 0

	if len(req.Nodes) == 0 {
		return DRCResponse{Errors: errors, Warnings: warnings, Passed: 0, Citations: Citations}
	}

	electrodes := electrodeNodes(req.Nodes)
	materials := materialNodes(req.Nodes)
	tissues := tissueNodes(req.Nodes)

	for _, electrode := range electrodes {
		local := design
		local.ElectrodeArea = firstPositive(numberValue(electrode.Data["area"]), design.ElectrodeArea)
		if material, ok := ResolveMaterial(lowerString(electrode.Data["material"])+" "+lowerString(electrode.Data["label"]), lowerString(electrode.Data["item_id"])); ok {
			local.Material = material
		}
		z, _ := randlesImpedance(1000, local.Material, local.ElectrodeArea)
		total := math.Abs(real(z))
		if total > maxImpedanceForSignal(local.Signal) {
			errors = append(errors, Violation{
				ID: "BIO-001", Title: "ELECTRODE IMPEDANCE OUT OF RANGE", Affected: labelOf(electrode),
				Message:  fmt.Sprintf("%s at %.0f um2 with %s is %.0f ohm at 1 kHz; %s sensing should stay below %.0f ohm or use a lower-impedance coating.", labelOf(electrode), local.ElectrodeArea, local.Material.Name, total, local.Signal.Name, maxImpedanceForSignal(local.Signal)),
				Severity: "error", Equation: "Z = Rs + ((Rct + Zw) || CPE)", Value: round(total, 1), Limit: maxImpedanceForSignal(local.Signal), Citation: "Randles 1947; Cogan 2008", Fixable: true,
				Autofix: map[string]any{"field": "area", "value": math.Ceil(local.ElectrodeArea * total / maxImpedanceForSignal(local.Signal))},
			})
		} else {
			passed++
		}

		if design.StimCurrentMA > 0 && local.Material.CIL > 0 {
			cd := chargeDensityMCPerCM2(local)
			if cd > local.Material.CIL {
				errors = append(errors, Violation{
					ID: "BIO-003", Title: "CHARGE INJECTION LIMIT EXCEEDED", Affected: labelOf(electrode),
					Message:  fmt.Sprintf("Charge density %.2f mC/cm2 exceeds %s limit %.2f mC/cm2. Reduce current, shorten pulse width, or increase area.", cd, local.Material.Name, local.Material.CIL),
					Severity: "error", Equation: "Qd = (I * pulse_width) / electrode_area", Value: round(cd, 2), Limit: local.Material.CIL, Citation: local.Material.Citation, Fixable: false,
				})
			} else {
				passed++
			}
		}
	}

	for _, materialNode := range materials {
		if material, ok := ResolveMaterial(lowerString(materialNode.Data["label"]), lowerString(materialNode.Data["item_id"])); ok {
			if !material.ISO10993 {
				errors = append(errors, Violation{
					ID: "BIO-006", Title: "MATERIAL NOT ISO 10993 APPROVED", Affected: labelOf(materialNode),
					Message:  fmt.Sprintf("%s is marked research-only in the curated library. Do not use it for in-vivo or investor-demo implant claims.", material.Name),
					Severity: "error", Citation: material.Citation, Fixable: false,
				})
			} else {
				passed++
			}
			if chronicTarget(tissues) && !material.ChronicSafe && material.CIL > 0 {
				warnings = append(warnings, Violation{
					ID: "WARN-007", Title: "MATERIAL NOT RATED FOR CHRONIC USE", Affected: labelOf(materialNode),
					Message:  fmt.Sprintf("%s is rated in this library for about %d year(s). Chronic designs should include drift/fouling margin or a stable coating.", material.Name, material.MaxYears),
					Severity: "warning", Citation: material.Citation, Fixable: false,
				})
			}
		}
	}

	if chronicTarget(tissues) && !design.HasEncapsulant {
		errors = append(errors, Violation{ID: "BIO-007", Title: "NO ENCAPSULANT ON CHRONIC IMPLANT", Affected: "Device Assembly", Message: "Chronic implant targets need a moisture barrier or housing such as Parylene-C or titanium.", Severity: "error", Citation: "ISO 10993 contact-duration framework; chronic implant packaging practice", Fixable: false})
	} else if chronicTarget(tissues) {
		passed++
	}

	if design.HasADC && !design.HasReference && len(electrodes) > 0 {
		warnings = append(warnings, Violation{ID: "BIO-004", Title: "NO REFERENCE ELECTRODE", Affected: "Measurement System", Message: "A stable reference electrode is needed for drift-controlled potentiometric/amperometric measurements.", Severity: "warning", Citation: "Electrochemical measurement practice", Fixable: false})
	} else if design.HasReference {
		passed++
	}

	if wearableTarget(tissues) && !design.HasFilter && len(electrodes) > 0 {
		warnings = append(warnings, Violation{ID: "WARN-003", Title: "NO MOTION ARTIFACT FILTER", Affected: "Signal Chain", Message: "Wearable tissue targets should include bandpass or adaptive motion rejection before SNR claims are trusted.", Severity: "warning", Equation: "noise_total = sqrt(sum(noise_i^2))", Citation: design.Tissue.Citation, Fixable: false})
	} else if design.HasFilter {
		passed++
	}

	if len(req.Nodes) > 4 && !design.HasMCU {
		warnings = append(warnings, Violation{ID: "WARN-020", Title: "NO DATA PROCESSING UNIT", Affected: "System Architecture", Message: "Complex designs need a controller, processor, or wireless/data path for a credible prototype flow.", Severity: "warning", Fixable: false})
	} else if design.HasMCU {
		passed++
	}

	if len(tissues) > 0 {
		passed++
	}
	if len(electrodes) > 0 {
		passed++
	}
	if len(req.Edges) > 0 {
		passed++
	}

	return DRCResponse{Errors: errors, Warnings: warnings, Passed: passed, Citations: Citations}
}

func maxImpedanceForSignal(signal SignalBand) float64 {
	switch signal.Key {
	case "spike":
		return 900000
	case "lfp", "eeg":
		return 250000
	case "glucose":
		return 5000000
	case "ecg":
		return 100000
	default:
		return 750000
	}
}

func electrodeNodes(nodes []Node) []Node {
	var out []Node
	for _, node := range nodes {
		label := lowerString(node.Data["label"])
		role := lowerString(node.Data["role"])
		nodeType := strings.ToLower(node.Type)
		if (nodeType == "electronics" || nodeType == "electrode" || nodeType == "neuromodulation") && 
		   (strings.Contains(label, "electrode") || strings.Contains(role, "interface") || strings.Contains(role, "electrode") || nodeType == "electrode") {
			out = append(out, node)
		}
	}
	return out
}

func materialNodes(nodes []Node) []Node {
	var out []Node
	for _, node := range nodes {
		if node.Type == "material" {
			out = append(out, node)
		}
	}
	return out
}

func tissueNodes(nodes []Node) []Node {
	var out []Node
	for _, node := range nodes {
		if node.Type == "biology" {
			out = append(out, node)
		}
	}
	return out
}

func chronicTarget(nodes []Node) bool {
	for _, node := range nodes {
		label := lowerString(node.Data["label"])
		if strings.Contains(label, "cortical") || strings.Contains(label, "nerve") || strings.Contains(label, "retina") || strings.Contains(label, "gastric") || strings.Contains(label, "chronic") {
			return true
		}
	}
	return false
}

func wearableTarget(nodes []Node) bool {
	for _, node := range nodes {
		label := lowerString(node.Data["label"])
		if strings.Contains(label, "skin") || strings.Contains(label, "subcutaneous") || strings.Contains(label, "wearable") {
			return true
		}
	}
	return false
}

func labelOf(node Node) string {
	label := toString(node.Data["label"])
	if label == "" {
		return node.ID
	}
	return label
}
